extensions/emmet/src/abbreviationActions.ts

```diff
@@ -216,66 +218,71 @@ function doWrapping(_: boolean, args: any) {
 		}, { undoStopBefore: false, undoStopAfter: false });
 	}
 
-	function makeChanges(inputAbbreviation: string | undefined, definitive: boolean): Thenable<boolean> {
+	let inPreviewMode = false;
+	async function makeChanges(inputAbbreviation: string | undefined, previewChanges: boolean): Promise<boolean> {
 		if (!inputAbbreviation || !inputAbbreviation.trim() || !helper.isAbbreviationValid(syntax, inputAbbreviation)) {
-			return inPreview ? revertPreview().then(() => { return false; }) : Promise.resolve(inPreview);
+			if (inPreviewMode) {
+				inPreviewMode = false;
+				await revertPreview();
+			}
+			return false;
 		}
 
 		const extractedResults = helper.extractAbbreviationFromText(inputAbbreviation);
 		if (!extractedResults) {
-			return Promise.resolve(inPreview);
-		} else if (extractedResults.abbreviation !== inputAbbreviation) {
-			// Not clear what should we do in this case. Warn the user? How?
+			if (inPreviewMode) {
+				inPreviewMode = false;
+				await revertPreview();
+			}
+			return false;
 		}
 
 		const { abbreviation, filter } = extractedResults;
-		if (definitive) {
-			const revertPromise = inPreview ? revertPreview() : Promise.resolve(true);
-			return revertPromise.then(() => {
-				const expandAbbrList: ExpandAbbreviationInput[] = rangesToReplace.map(rangesAndContent => {
-					const rangeToReplace = rangesAndContent.originalRange;
-					let textToWrap: string[];
-					// if (individualLines) {
-					textToWrap = rangesAndContent.textToWrapInPreview;
-					// } else {
-					// 	// use the p tag as a dummy element to get Emmet to wrap the expression properly
-					// 	textToWrap = rangeToReplace.isSingleLine ?
-					// 		['$TM_SELECTED_TEXT'] : ['<p>$TM_SELECTED_TEXT</p>'];
-					// }
-					return { syntax: syntax || '', abbreviation, rangeToReplace, textToWrap, filter };
-				});
-				return expandAbbreviationInRange(editor, expandAbbrList, false).then(() => { return true; });
-			});
+		if (abbreviation !== inputAbbreviation) {
+			// Not clear what should we do in this case. Warn the user? How?
 		}
 
-		const expandAbbrList: ExpandAbbreviationInput[] = rangesToReplace.map(rangesAndContent => {
-			return { syntax: syntax || '', abbreviation, rangeToReplace: rangesAndContent.originalRange, textToWrap: rangesAndContent.textToWrapInPreview, filter };
-		});
+		if (previewChanges) {
+			const expandAbbrList: ExpandAbbreviationInput[] = rangesToReplace.map(rangesAndContent =>
+				({ syntax, abbreviation, rangeToReplace: rangesAndContent.originalRange, textToWrap: rangesAndContent.textToWrapInPreview, filter, indent, baseIndent: rangesAndContent.baseIndent })
+			);
+
+			inPreviewMode = true;
+			return applyPreview(expandAbbrList);
+		}
+
+		const expandAbbrList: ExpandAbbreviationInput[] = rangesToReplace.map(rangesAndContent =>
+			({ syntax, abbreviation, rangeToReplace: rangesAndContent.originalRange, textToWrap: rangesAndContent.textToWrapInPreview, filter, indent })
+		);
+
+		if (inPreviewMode) {
```

Comment:
This if block shows up a few times and seems repetitive. Wondering whether it can be put into its own async function.

Reply:
Do you mean moving the `if (inPreviewMode)` check inside `revertPreview` function?
Could be but I think it's better/readable to have just one place (`makeChanges` function) were the `inPreviewMode` variable is modified than having multiple places.